src/test/java/org/jenkinsci/plugins/workflow/job/WorkflowRunTest.java

```diff
@@ -667,19 +685,42 @@ public void onInitialize(Run run) {
 
     }
 
-
-    @SafeVarargs
-    private static void assertPollingBaselines(List<WorkflowRun.SCMCheckout> checkouts, Matcher<Object>... indexedMatchers) {
-        assertThat("Number of checkouts should match number of matchers", checkouts.size(), equalTo(indexedMatchers.length));
-        for (int i = 0; i < checkouts.size(); i++) {
-            assertThat("Unexpected baseline for checkout at index " + i, checkouts.get(i).pollingBaseline, indexedMatchers[i]);
+    @Issue("JENKINS-73835")
+    @Test public void logRotationOnlyProcessesCompletedBuilds() throws Throwable {
+        logging.record(LogRotator.class, Level.FINER);
+        var p = r.createProject(WorkflowJob.class);
+        p.setDefinition(new CpsFlowDefinition(
+                "echo params.FOO; semaphore 'wait'", true));
+        p.addProperty(new ParametersDefinitionProperty(List.of(new StringParameterDefinition("FOO"))));
+        // Keep 0 builds, i.e. delete all builds immediately.
+        LogRotator logRotator = new LogRotator(-1, 0, -1, -1);
+        logRotator.setRemoveLastBuild(true);
+        p.setBuildDiscarder(logRotator);
+        int buildsToRun = 10; // Increase this number to reproduce the issue more easily prior to the fix.
+        Run[] builds = new Run[buildsToRun];
+        File[] buildDirs = new File[buildsToRun];
+        // Run a large number of builds that should finish around the same time to check race conditions with log rotation and build completion.
+        for (int i = 0; i < buildsToRun; i++) {
+            var b = p.scheduleBuild2(0, new ParametersAction(List.of(new StringParameterValue("FOO", "b" + i)))).waitForStart();
+            builds[i] = b;
+            buildDirs[i] = b.getRootDir();
+            SemaphoreStep.waitForStart("wait/" + (i+1), b);
+        }
+        for (int i = 0; i < buildsToRun; i++) {
+            SemaphoreStep.success("wait/" + (i+1), null);
+        }
+        LOGGER.info("Waiting for all builds to complete");
+        for (int i = 0; i < buildsToRun; i++) {
+            r.waitForCompletion(builds[i]);
+        }
+        LOGGER.info("Checking that all build directories are empty ");
+        for (int i = 0; i < buildsToRun; i++) {
+            String[] filesInBuildDir = buildDirs[i].list();
+            if (filesInBuildDir == null) {
+                filesInBuildDir = new String[0];
+            }
+            assertThat("Expected " + buildDirs[i] + " to be empty but saw: " + Arrays.toString(filesInBuildDir), filesInBuildDir, emptyArray());
```

Comment:
Prior to the fix, this fails intermittently because some directory contains `build.xml`.